### PR TITLE
fix(distributed): synchronize convergence across ranks

### PR DESCRIPTION
**Pull request by atulcthakur**
# ALCHEMI Toolkit Pull Request

## Description

`DomainParallel` treated convergence indices as Boolean values. Because `tensor([0])` means “system 0 converged” but converts to `False`, convergence could be missed. 
Ranks could also disagree locally and enter different collectives, causing a hang. 
The DD run and pipeline paths also failed to stop or retire the converged system.

## Fix

- Preserve convergence as system-index tensors.
- Reduce a per-system convergence mask across every domain rank.
- Stop completed DD runs and retire converged pipeline systems.
- Avoid convergence collectives when no convergence hook is configured.

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

## Changes Made

- `nvalchemi/distributed/domain_parallel.py`
  - Fix distributed convergence reduction in `step()`.
  - Add convergence-based early exit to `run()`.
  - Fix pipeline completion and completion-reason checks.

- `test/distributed/test_domain_parallel_convergence.py`
  - Test preservation of `tensor([0])`.
  - Test rank disagreement without hanging.
  - Test that fixed-step dynamics adds no convergence collective.
  - Test `run()` early exit.
  - Test pipeline-stage retirement.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [ ] New tests added for new functionality meets coverage expectations?

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [ ] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [ ] I have added docstrings to new functions/classes
- [ ] I have updated the documentation (if applicable)

## Additional Notes

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

**Comment by atulcthakur** [comment]
/ok to test https://github.com/NVIDIA/nvalchemi-toolkit/pull/150/commits/6a34c19b06a1c15e808a55b298fbc07a9ea239f0

File: nvalchemi/distributed/domain_parallel.py [hunk 1]
```diff
@@ -326,22 +326,30 @@ def step(self, batch: Batch) -> tuple[Batch, torch.Tensor | None]:
...
         if (
-            converged is not None
+            dyn.convergence_hook is not None
             and dist.is_initialized()
             and self._config.mesh is not None
         ):
-            flag = torch.tensor(
-                [1 if bool(converged) else 0],
+            convergence_mask = torch.zeros(
+                batch.num_graphs,
```
Comment (zubatyuk) [inline nvalchemi/distributed/domain_parallel.py:339]: `ShardedBatch.num_graphs` is always 1

File: nvalchemi/distributed/domain_parallel.py [hunk 2]
```diff
@@ -326,22 +326,30 @@ def step(self, batch: Batch) -> tuple[Batch, torch.Tensor | None]:
...
             )
+            if converged is not None:
+                local_indices = converged.to(
+                    device=convergence_mask.device, dtype=torch.long
+                )
+                convergence_mask[local_indices] = 1
             dist.all_reduce(
-                flag, op=dist.ReduceOp.MIN, group=mesh_group(self._config.mesh)
+                convergence_mask,
+                op=dist.ReduceOp.MIN,
```
Comment (zubatyuk) [inline nvalchemi/distributed/domain_parallel.py:350]: This works for MAX convergence criteria. Should RMS (like RMS force) from custom op be supported as well?

File: test/distributed/test_domain_parallel_convergence.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    dist.barrier()
+    queue.put((rank, _serialize_convergence(converged)))
+
+
+def test_converged_system_zero_remains_an_index_tensor() -> None:
+    """``tensor([0])`` means system zero converged; zero is not ``False``."""
+    results = sorted(
+        run_gloo(world_size=2, fn=_all_ranks_converged_worker),
+        key=lambda item: item[0],
+    )
```
Comment (zubatyuk) [inline test/distributed/test_domain_parallel_convergence.py:321]: These workers give every rank a complete artificial batch and mock partitioning, model execution, and integration. Could we add one two-rank regression that starts with a full system on rank 0, calls partition(), evaluates convergence on genuinely different owned atoms, and drives run() until it exits? Comparing the result with the full-system convergence calculation would cover ownership, the collective, and early termination together.

File: test/distributed/test_domain_parallel_convergence.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    ):
+        _, converged = dp.step(batch)
+
+    assert converged is None
+    all_reduce.assert_not_called()
+
+
+def test_run_stops_after_every_system_has_converged() -> None:
+    """A distributed run should stop once all resident systems converge."""
+    dp = _single_process_domain_parallel()
```
Comment (zubatyuk) [inline test/distributed/test_domain_parallel_convergence.py:434]: This calls `_poststep_sync_buffers()` directly on an isolated stage, so it does not verify retirement through `DistributedPipeline.run()` or agreement across the domain group. Could we add a small grouped-pipeline test that reaches convergence through the normal driver and confirms that every rank retires the system and terminates?

File: nvalchemi/distributed/domain_parallel.py [hunk 5]
```diff
@@ -326,22 +326,30 @@ def step(self, batch: Batch) -> tuple[Batch, torch.Tensor | None]:
...
+                group=mesh_group(self._config.mesh),
```
Comment (zubatyuk) [inline nvalchemi/distributed/domain_parallel.py]: Could this use `self._strategy.process_group`? `mesh_group(self._config.mesh)` ignores mesh_dim, so this fails for a multidimensional mesh.

File: nvalchemi/distributed/domain_parallel.py [hunk 6]
```diff
@@ -326,22 +326,30 @@ def step(self, batch: Batch) -> tuple[Batch, torch.Tensor | None]:
         converged = dyn._check_convergence(batch)
         # Convergence must be a mesh-wide decision: each rank only sees its own
         # atoms, so ranks can disagree and take divergent control flow (one stops
-        # while others continue → collective desync / hang). Reduce to the AND
-        # across the domain — converged only when EVERY rank is converged.
+        # while others continue → collective desync / hang). Every rank reduces
+        # a per-system mask so an index survives only when every rank reports it.
         if (
-            converged is not None
+            dyn.convergence_hook is not None
```
Comment (zubatyuk) [inline nvalchemi/distributed/domain_parallel.py]: Should honor `dyn.convergence_hook.frequency`

File: nvalchemi/distributed/domain_parallel.py (hunk 1 above)
Comment (laserkelvin) [reply nvalchemi/distributed/domain_parallel.py:339]: I'm fine with the modification because it's API consistent, but add an inline comment to say that it's always 1

File: nvalchemi/distributed/domain_parallel.py (hunk 2 above)
Comment (laserkelvin) [reply nvalchemi/distributed/domain_parallel.py:350]: This is a good to have I think, but given that the scope changes a fair amount, I would suggest we defer supporting RMS convergence to the next release.

File: nvalchemi/distributed/domain_parallel.py (hunk 6 above)
Comment (atulcthakur) [reply nvalchemi/distributed/domain_parallel.py]: Fixed. Thanks.

File: nvalchemi/distributed/domain_parallel.py (hunk 5 above)
Comment (atulcthakur) [reply nvalchemi/distributed/domain_parallel.py]: Fixed. Thanks.

File: test/distributed/test_domain_parallel_convergence.py (hunk 3 above)
Comment (atulcthakur) [reply test/distributed/test_domain_parallel_convergence.py:321]: Added a two-rank test using real `DomainParallel.step()` and `run()`

File: test/distributed/test_domain_parallel_convergence.py (hunk 4 above)
Comment (atulcthakur) [reply test/distributed/test_domain_parallel_convergence.py:434]: Added a four-rank test using a 2 pipeline × 2 domain mesh. It runs the normal `DistributedPipeline.run()` path and verifies stage retirement.

File: nvalchemi/distributed/domain_parallel.py (hunk 2 above)
Comment (atulcthakur) [reply nvalchemi/distributed/domain_parallel.py:350]: Agree. Keeping the changes to min required for now. 
Marked for follow up.

File: nvalchemi/distributed/domain_parallel.py (hunk 1 above)
Comment (atulcthakur) [reply nvalchemi/distributed/domain_parallel.py:339]: Added a comment. We split only 1 system across ranks for now.

File: nvalchemi/distributed/domain_parallel.py [hunk 7]
```diff
@@ -715,7 +729,11 @@ def _bcast_group_flag(self, flag: bool) -> bool:
...
         """A resident system leaves this stage when it converges (FIRE) or spends
         its per-system step budget (``n_steps``, e.g. an NVT leg)."""
-        if converged:
+        if (
+            converged is not None
+            and self.active_batch is not None
+            and converged.numel() == self.active_batch.num_graphs
+        ):
             return True
         return self.n_steps is not None and self._system_step >= self.n_steps
```
[the remark this answers is not in the document]
Comment (atulcthakur) [reply nvalchemi/distributed/domain_parallel.py]: Fixed.

**Comment by atulcthakur** [comment]
/ok to test https://github.com/NVIDIA/nvalchemi-toolkit/pull/150/commits/63c61614fc78aa90b9beabe188e3cc7e832ce1af

**Comment by atulcthakur** [comment]
/ok to test https://github.com/NVIDIA/nvalchemi-toolkit/pull/150/commits/369a08b7f3d3ea79b349b941e76d66a2f77c0a7f

**Comment by atulcthakur** [comment]
Marking this as ready to merge from my end
